pkg/rpc/server/server.go

```diff
@@ -1873,7 +1889,153 @@ func (s *Server) runScriptInVM(t trigger.Type, script []byte, contractScriptHash
 	if err != nil {
 		faultException = err.Error()
 	}
-	return result.NewInvoke(ic, script, faultException, s.config.MaxIteratorResultItems), nil
+	var registerSession result.OnNewSession
+	if s.config.SessionEnabled {
+		registerSession = s.registerSession
+	}
+	return result.NewInvoke(ic, script, faultException, registerSession, s.config.MaxIteratorResultItems), nil
+}
+
+// registerSession is a callback used to add new iterator session to the sessions list.
+// It performs no check whether sessions are enabled.
+func (s *Server) registerSession(sessionID string, iterators []result.ServerIterator, finalize func()) {
+	s.sessionsLock.Lock()
+	sess := &session{
+		iterators: iterators,
+		finalize:  finalize,
+		till:      time.Now().Add(time.Second * time.Duration(s.config.SessionExpirationTime)),
+	}
+	s.sessions[sessionID] = sess
+	s.sessionsLock.Unlock()
+}
+
+func (s *Server) handleSessions() {
+	if !s.config.SessionEnabled {
+		return
+	}
+	s.log.Info("rpc-server iterator sessions are enabled")
+	var (
+		finalizationInterval = time.Duration(s.config.SessionExpirationTime) * time.Second
+		finalizationTicker   = time.NewTicker(finalizationInterval)
+	)
+
+	defer func() {
+		finalizationTicker.Stop()
+		s.sessionsLock.Lock()
+		defer s.sessionsLock.Unlock()
+		for _, session := range s.sessions {
+			// Concurrent iterator traversal may still be in process, thus need to protect iterators access.
+			session.iteratorsLock.Lock()
+			session.finalize()
+			session.iteratorsLock.Unlock()
+		}
+		s.sessions = nil
+	}()
+
+	for {
+		select {
+		case <-s.shutdown:
+			return
+		case <-finalizationTicker.C:
+			s.sessionsLock.Lock()
+			if len(s.sessions) == 0 {
+				s.sessionsLock.Unlock()
+				continue
+			}
+			now := time.Now()
+			staleSessions := s.sessions
+			// We can easily reuse existing map, but the ticker has tick period equal to the session expiration, so likely,
+			// the whole set of sessions will be removed from map at the subsequent tick, it's cheaper than using `delete`
+			// for all set of elements.
```

Comment:
Actually, a half is to be removed if we're to assume normal request distribution.

Reply:
You're right. So is it better to use `delete` on existing map?

Reply:
With `AfterFunc` we'll need `delete` anyway.